src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -616,14 +614,17 @@ public byte[] genClassBytes(Configuration cf) {
                         // generate <init>
                         genConstructor(clb);
 
+                        // generate dedup set fields and provider methods
+                        var dedupSets = genConstants(clb);
+
```

Comment:
These constants are all for building the module descriptors.   This can be moved and have `genModuleDescriptorsMethod` to call `genConstants`.

Reply:
The static initializer currently only contains the cache from DedupSet, but it feels wrong to generate the static initializer in genModuleDescriptorsMethod.

Reply:
I was wondering that there should be `genClassInitializer`.   It seems to me that `SystemModulesClassGenerator` should have `List<Snippet> clinitSnippets;` and `genClassInitializer` will emit the code from the snippets if not empty.   It'd have to be called at the end since other code might add clinit snippet in the future.